# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [59]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
load_dotenv()
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/jcmignon/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [60]:
# Read your api key environment variable
api_key = os.getenv("NEWS_API_KEY")

In [61]:
# Create a newsapi client
from newsapi import NewsApiClient
newsapi = NewsApiClient(api_key=api_key)

In [62]:
# Fetch the Bitcoin news articles
btc_news1 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 1)['articles'])
btc_news2 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 2)['articles'])
btc_news3 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 3)['articles'])
btc_news4 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 4)['articles'])
btc_news5 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 5)['articles'])

btc_news = pd.concat([btc_news1,btc_news2,btc_news3,btc_news4,btc_news5])
btc_news

source  \
0   {'id': 'engadget', 'name': 'Engadget'}   
1      {'id': None, 'name': 'Gizmodo.com'}   
2   {'id': 'mashable', 'name': 'Mashable'}   
3   {'id': 'bbc-news', 'name': 'BBC News'}   
4      {'id': None, 'name': 'Gizmodo.com'}   
..                                     ...   
15    {'id': 'reuters', 'name': 'Reuters'}   
16    {'id': 'reuters', 'name': 'Reuters'}   
17    {'id': 'reuters', 'name': 'Reuters'}   
18    {'id': 'reuters', 'name': 'Reuters'}   
19    {'id': 'reuters', 'name': 'Reuters'}   

                                               author  \
0   https://www.engadget.com/about/editors/saqib-shah   
1                                          Matt Novak   
2                                      Stan Schroeder   
3                    https://www.facebook.com/bbcnews   
4                                       Alyse Stanley   
..                                                ...   
15                                            Reuters   
16                                         Tom Wilson   
17                                      Reuters Staff   
18                           Gertrude Chavez-dreyfuss   
19                                            Reuters   

                                                title  \
0   El Salvador becomes the first country to appro...   
1   El Salvador Becomes First Country to Recognize...   
2   Elon Musk says Tesla will resume Bitcoin purch...   
3   Bitcoin: El Salvador makes cryptocurrency lega...   
4   Miami's Bitcoin Conference May Be the Latest C...   
..                                                ...   
15  Square to invest $5 mln in Blockstream's solar...   
16  EXCLUSIVE El Salvador bitcoin transfers soar, ...   
17  IMF sees legal, economic issues with El Salvad...   
18  Crypto sees 2nd week of outflows; ether posts ...   
19  IMF sees legal, economic issues with El Salvad...   

                                          description  \
0   El Salvador has voted to adopt Bitcoin as lega...   
1   El Salvador has become the first country in th...   
2   It's all about clean energy, it seems. \nElon ...   
3   It is the first country in the world to make t...   
4   Several crypto fans that descended on Miami, F...   
..                                                ...   
15  Blockchain technology company Blockstream Mini...   
16  Small transfers of bitcoin to El Salvador jump...   
17  The International Monetary Fund said on Thursd...   
18  Cryptocurrency investment products and funds s...   
19  The International Monetary Fund said on Thursd...   

                                                  url  \
0   https://www.engadget.com/el-salvador-bitcoin-l...   
1   https://gizmodo.com/el-salvador-becomes-first-...   
2   https://mashable.com/article/tesla-bitcoin-pur...   
3   https://www.bbc.co.uk/news/world-latin-america...   
4   https://gizmodo.com/miamis-bitcoin-conference-...   
..                                                ...   
15  https://www.reuters.com/technology/square-inve...   
16  https://www.reuters.com/business/finance/exclu...   
17  https://www.reuters.com/article/us-el-salvador...   
18  https://www.reuters.com/technology/crypto-sees...   
19  https://www.reuters.com/business/finance/imf-s...   

                                           urlToImage           publishedAt  \
0   https://s.yimg.com/os/creatr-uploaded-images/2...  2021-06-09T12:04:40Z   
1   https://i.kinja-img.com/gawker-media/image/upl...  2021-06-09T10:00:00Z   
2   https://mondrian.mashable.com/2021%252F06%252F...  2021-06-14T07:15:49Z   
3   https://ichef.bbci.co.uk/news/1024/branded_new...  2021-06-09T08:27:58Z   
4   https://i.kinja-img.com/gawker-media/image/upl...  2021-06-11T00:45:00Z   
..                                                ...                   ...   
15                                               None  2021-06-05T18:22:00Z   
16  https://www.reuters.com/resizer/AUSGmfvZnVBr2N...  2021-06-14T14:24:00Z   
17  https://static.reuters.com/r

In [63]:
# Fetch the Ethereum news articles
eth_news = newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy')

eth_news1 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 1)['articles'])
eth_news2 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 2)['articles'])
eth_news3 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 3)['articles'])
eth_news4 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 4)['articles'])
eth_news5 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 5)['articles'])

eth_news = pd.concat([eth_news1,eth_news2,eth_news3,eth_news4,eth_news5])
eth_news.head()

source  \
0             {'id': 'mashable', 'name': 'Mashable'}   
1             {'id': 'mashable', 'name': 'Mashable'}   
2               {'id': None, 'name': 'Entrepreneur'}   
3  {'id': 'business-insider', 'name': 'Business I...   
4                {'id': None, 'name': 'Gizmodo.com'}   

                                   author  \
0                            Joseph Green   
1                              Tim Marcin   
2                 Entrepreneur en Español   
3  cshumba@insider.com (Camomile Shumba )   
4                         Whitney Kimball   

                                               title  \
0  This blockchain development course bundle is o...   
1               Classic memes that have sold as NFTs   
2  Ethereum creator Vitalik Buterin made more tha...   
3  Ethereum now has more active addresses than bi...   
4                                    8 Money Toilets   

                                         description  \
0  TL;DR: The Cryptocurrency with Ethereum and So...   
1  It wasn't long ago that your average person ha...   
2  Ethereum billionaire founder Vitalik Buterin b...   
3  Ethereum overtook bitcoin in the number of act...   
4  A ferocious bidding war over a misshapen chick...   

                                                 url  \
0  https://mashable.com/uk/shopping/june-17-crypt...   
1  https://mashable.com/article/classic-memes-sol...   
2        https://www.entrepreneur.com/article/374264   
3  https://markets.businessinsider.com/news/stock...   
4     https://gizmodo.com/8-money-toilets-1847036893   

                                          urlToImage           publishedAt  \
0  https://mondrian.mashable.com/2021%252F06%252F...  2021-06-17T04:05:00Z   
1  https://mondrian.mashable.com/2021%252F06%252F...  2021-06-20T19:28:07Z   
2  https://assets.entrepreneur.com/content/3x2/20...  2021-06-10T19:12:00Z   
3  https://images2.markets.businessinsider.com/60...  2021-07-02T15:49:56Z   
4  https://i.kinja-img.com/gawker-media/image/upl...  2021-06-04T22:20:00Z   

                                             content  
0  TL;DR: The Cryptocurrency with Ethereum and So...  
1  It wasn't long ago that your average person ha...  
2  This article was translated from our Spanish e...  
3  Bitcoin and Ethereum\r\nYuriko Nakao\r\nEther ...  
4  CryptoPunks represent the only historically re...

In [64]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiment = []

for content in btc_news['content']:
    sentiment = analyzer.polarity_scores(content)
    compound = sentiment["compound"]
    pos = sentiment["pos"]
    neu = sentiment["neu"]
    neg = sentiment["neg"]
        
    btc_sentiment.append({
            "text": content,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu   
        })

# Create DataFrame
btc_df = pd.DataFrame(btc_sentiment)

# Reorder DataFrame columns
cols = ["text", "compound", "positive", "negative", "neutral"]
btc_df = btc_df[cols]

btc_df.head()

text  compound  positive  \
0  El Salvador's President Nayib Bukele has made ...    0.8402     0.282   
1  El Salvador has become the first country in th...    0.1280     0.043   
2  It's all about clean energy, it seems. \r\nElo...    0.6908     0.169   
3  image captionThe move means bitcoin will be ac...    0.2732     0.060   
4  Several crypto fans that descended on Miami, F...    0.5574     0.107   

   negative  neutral  
0       0.0    0.718  
1       0.0    0.957  
2       0.0    0.831  
3       0.0    0.940  
4       0.0    0.893

In [65]:
# Create the Bitcoin sentiment scores DataFrame
eth_sentiment = []

for content in eth_news['content']:
    sentiment = analyzer.polarity_scores(content)
    compound = sentiment["compound"]
    pos = sentiment["pos"]
    neu = sentiment["neu"]
    neg = sentiment["neg"]
        
    eth_sentiment.append({
            "text": content,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu   
        })

# Create DataFrame
eth_df = pd.DataFrame(eth_sentiment)

# Reorder DataFrame columns
cols = ["text", "compound", "positive", "negative", "neutral"]
eth_df = eth_df[cols]

eth_df.head()

text  compound  positive  \
0  TL;DR: The Cryptocurrency with Ethereum and So...    0.0000     0.000   
1  It wasn't long ago that your average person ha...   -0.2960     0.000   
2  This article was translated from our Spanish e...   -0.3400     0.000   
3  Bitcoin and Ethereum\r\nYuriko Nakao\r\nEther ...    0.3612     0.110   
4  CryptoPunks represent the only historically re...   -0.4588     0.067   

   negative  neutral  
0     0.000    1.000  
1     0.061    0.939  
2     0.066    0.934  
3     0.041    0.849  
4     0.151    0.782

In [66]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound    positive    negative     neutral
count  100.000000  100.000000  100.000000  100.000000
mean     0.061507    0.047040    0.028200    0.924750
std      0.352836    0.063612    0.043421    0.070162
min     -0.778300    0.000000    0.000000    0.718000
25%     -0.081925    0.000000    0.000000    0.871000
50%      0.000000    0.000000    0.000000    0.929000
75%      0.323650    0.084000    0.060500    1.000000
max      0.840200    0.282000    0.178000    1.000000

In [67]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound   positive    negative     neutral
count  100.000000  100.00000  100.000000  100.000000
mean     0.150449    0.05848    0.022150    0.919390
std      0.362227    0.06918    0.036668    0.071027
min     -0.571900    0.00000    0.000000    0.710000
25%      0.000000    0.00000    0.000000    0.875250
50%      0.051350    0.04600    0.000000    0.927000
75%      0.411525    0.09725    0.057750    1.000000
max      0.848100    0.29000    0.151000    1.000000

### Questions:

Q: Which coin had the highest mean positive score?

*__A: Etheruem had the highest mean positive score, 5.49% vs 4.68% for Bitcoin.__*

Q: Which coin had the highest compound score?

*__A: Ethereum had the highest mean compound score (12.84% vs 5.92% respectively), as well as the highest compound score for an individual article (84.81% vs 84.02% respectively).__*

Q. Which coin had the highest positive score?

*__A: Ethereum had the highest positive score (29% vs 28.2% for ETH)__*

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [68]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import reuters, stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [69]:
btc_string=" ".join(btc_news['content'].astype(str))

In [70]:
# Create a list of stopwords
sentence = sent_tokenize(btc_string)
words = word_tokenize(btc_string)
sw = set(stopwords.words('english'))
first_result = [word.lower() for word in words if word.lower() not in sw]

# unsure if the list has to be expanded 
# Expand list with char reuters

In [71]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

In [72]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', text)

    # Create a tokenized list of the words
    words = word_tokenize(re_clean)    
    
    # Define the stop words
    sw = stopwords.words('english')
    # Additional words as they appear in the top 10 most common words and add little to no value
    sw_addon = {'char', 'reuters'}
    
    # Lemmatize words into root words
    lem = [lemmatizer.lemmatize(word) for word in words]
   
    # Convert the words to lowercase
    output = [word.lower() for word in lem if word.lower() not in sw]
    
    # unsure how to add the custom list - .union(sw_addon) - to be pasted behind sw. in the previous line, error: union is not recognized
    
    return output

In [73]:
# Create a new tokens column for Bitcoin
btc_token = btc_news['content'].apply(tokenizer)
btc_df['token'] = btc_token.values
btc_df.head()

text  compound  positive  \
0  El Salvador's President Nayib Bukele has made ...    0.8402     0.282   
1  El Salvador has become the first country in th...    0.1280     0.043   
2  It's all about clean energy, it seems. \r\nElo...    0.6908     0.169   
3  image captionThe move means bitcoin will be ac...    0.2732     0.060   
4  Several crypto fans that descended on Miami, F...    0.5574     0.107   

   negative  neutral                                              token  
0       0.0    0.718  [el, salvadors, president, nayib, bukele, ha, ...  
1       0.0    0.957  [el, salvador, ha, become, first, country, wor...  
2       0.0    0.831  [clean, energy, seemselon, musk, tesla, caused...  
3       0.0    0.940  [image, captionthe, move, mean, bitcoin, accep...  
4       0.0    0.893  [several, crypto, fan, descended, miami, flori...

In [74]:
# Create a new tokens column for Ethereum
eth_token = eth_news['content'].apply(tokenizer)
eth_df['token'] = eth_token.values
eth_df.head()

text  compound  positive  \
0  TL;DR: The Cryptocurrency with Ethereum and So...    0.0000     0.000   
1  It wasn't long ago that your average person ha...   -0.2960     0.000   
2  This article was translated from our Spanish e...   -0.3400     0.000   
3  Bitcoin and Ethereum\r\nYuriko Nakao\r\nEther ...    0.3612     0.110   
4  CryptoPunks represent the only historically re...   -0.4588     0.067   

   negative  neutral                                              token  
0     0.000    1.000  [tldr, cryptocurrency, ethereum, solidity, blo...  
1     0.061    0.939  [wasnt, long, ago, average, person, clue, nft,...  
2     0.066    0.934  [article, wa, translated, spanish, edition, us...  
3     0.041    0.849  [bitcoin, ethereumyuriko, nakaoether, overtook...  
4     0.151    0.782  [cryptopunks, represent, historically, relevan...

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [75]:
from collections import Counter
from nltk import ngrams

In [76]:
# Create btc_string of all tokenized words per article 
btc_string = " ".join(btc_df['token'].astype(str))
# Tokenize btc_string to ensure you have one big string 
btc_string_processed = tokenizer(btc_string)

In [77]:
# Generate the Bitcoin N-grams where N=2
btc_counts = Counter(ngrams(btc_string_processed, n=2))
print(dict(btc_counts))

{('el', 'salvador'): 21, ('salvador', 'president'): 2, ('president', 'nayib'): 7, ('nayib', 'bukele'): 6, ('bukele', 'ha'): 1, ('ha', 'made'): 1, ('made', 'good'): 1, ('good', 'promise'): 1, ('promise', 'adopt'): 1, ('adopt', 'bitcoin'): 3, ('bitcoin', 'legal'): 12, ('legal', 'tender'): 11, ('tender', 'official'): 1, ('official', 'central'): 1, ('central', 'american'): 3, ('american', 'country'): 1, ('country', 'congress'): 1, ('congress', 'voted'): 1, ('voted', 'accept'): 1, ('accept', 'cryptocurrency'): 1, ('cryptocurrency', 'majori'): 1, ('majori', 'char'): 1, ('char', 'el'): 5, ('salvador', 'ha'): 6, ('ha', 'become'): 6, ('become', 'first'): 5, ('first', 'country'): 6, ('country', 'world'): 6, ('world', 'recognize'): 1, ('recognize', 'cryptocurrency'): 1, ('cryptocurrency', 'bitcoin'): 4, ('legal', 'currency'): 1, ('currency', 'according'): 1, ('according', 'president'): 1, ('bukele', 'tweet'): 1, ('tweet', 'wednesday'): 1, ('wednesday', 'citizen'): 1, ('citizen', 'able'): 1, ('abl

In [78]:
# Generate the Ethereum N-grams where N=2
# Create eth_string of all tokenized words per article 
eth_string = " ".join(eth_df['token'].astype(str))
# Tokenize eth_string to ensure you have one big string 
eth_string_processed = tokenizer(eth_string)
# Generate the Ethereum N-grams where N=2
eth_counts = Counter(ngrams(eth_string_processed, n=2))
print(dict(eth_counts))

{('tldr', 'cryptocurrency'): 2, ('cryptocurrency', 'ethereum'): 4, ('ethereum', 'solidity'): 2, ('solidity', 'blockchain'): 2, ('blockchain', 'developer'): 2, ('developer', 'bundle'): 2, ('bundle', 'sale'): 1, ('sale', 'june'): 1, ('june', 'saving'): 1, ('saving', 'list'): 1, ('list', 'priceis'): 1, ('priceis', 'everyone'): 1, ('everyone', 'know'): 1, ('know', 'investing'): 1, ('investing', 'cryptocurrency'): 1, ('cryptocurrency', 'char'): 9, ('char', 'wasnt'): 1, ('wasnt', 'long'): 2, ('long', 'ago'): 2, ('ago', 'average'): 1, ('average', 'person'): 1, ('person', 'clue'): 1, ('clue', 'nft'): 1, ('nft', 'wa'): 1, ('wa', 'nowadays'): 1, ('nowadays', 'theyre'): 1, ('theyre', 'making'): 1, ('making', 'headline'): 1, ('headline', 'selling'): 1, ('selling', 'hundred'): 1, ('hundred', 'thousand'): 1, ('thousand', 'buck'): 1, ('buck', 'predictably'): 1, ('predictably', 'folk'): 1, ('folk', 'learning'): 1, ('learning', 'nfts'): 1, ('nfts', 'char'): 1, ('char', 'article'): 5, ('article', 'wa'):

In [79]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=10):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [80]:
# Use token_count to get the top 10 words for Bitcoin

# Create string with individual words
btc_counts_indiv = Counter(ngrams(btc_string_processed, n=1))
# Apply token_count function
token_count(btc_counts_indiv)

#Alternative solution
# print(dict(btc_counts_indiv.most_common(10)))

[(('char',), 97),
 (('bitcoin',), 96),
 (('reuters',), 57),
 (('june',), 38),
 (('cryptocurrency',), 31),
 (('salvador',), 28),
 (('el',), 26),
 (('world',), 24),
 (('monday',), 19),
 (('seen',), 19)]

In [35]:
# Use token_count to get the top 10 words for Ethereum

# Create string with individual words
eth_counts_indiv = Counter(ngrams(eth_string_processed, n=1))
# Apply token_count function
token_count(eth_counts_indiv)

#Alternative solution
# print(dict(eth_counts_indiv.most_common(10)))

[(('char',), 99),
 (('bitcoin',), 41),
 (('reuters',), 34),
 (('cryptocurrency',), 33),
 (('world',), 26),
 (('june',), 23),
 (('ethereum',), 14),
 (('biggest',), 14),
 (('representation',), 14),
 (('week',), 13)]

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [81]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [82]:
# Generate the Bitcoin word cloud

# confused - btc_string_processed is already a string (error given - expects a string like object)

wc = WordCloud().generate(btc_string_processed)
plt.imshow(wc)

TypeError: expected string or bytes-like object

In [23]:
# Generate the Ethereum word cloud

wc = WordCloud().generate(eth_string_processed)
plt.imshow(wc)

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [89]:
import spacy
from spacy import displacy

In [91]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [92]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [103]:
# Concatenate all of the Bitcoin text together
#already done in btc_string

In [96]:
# Run the NER processor on all of the text
btc_NER = nlp(btc_string)

# Add a title to the document
# YOUR CODE HERE!

In [97]:
# Render the visualization
displacy.render(btc_NER, style='ent')

In [101]:
# List all Entities
# all geopolitical entities like in the example. Not sure if that's the question
print([ent.text for ent in btc_NER.ents if ent.label_ == 'GPE'])

['florida', 'china', 'germany', 'china', "'york'", 'china', 'china', 'paris', "'salvador'", "'york'", "'york'", 'london', "'york'", "'york'", 'china', 'america', "'york'", 'london', 'china', "'beach'", 'washington', "'york'"]


---

### Ethereum NER

In [104]:
# Concatenate all of the Ethereum text together
# already done in eth_string

In [105]:
# Run the NER processor on all of the text
eth_NER = nlp(eth_string)

# Add a title to the document
# YOUR CODE HERE!

In [106]:
# Render the visualization
displacy.render(eth_NER, style='ent')

In [107]:
# List all Entities
# all geopolitical entities like in the example. Not sure if that's the question
print([ent.text for ent in eth_NER.ents if ent.label_ == 'GPE'])

["'york'", 'china', 'london', 'britain', 'london', 'london', 'staffmumbai', "'york'", 'london', 'london', 'shanghai', 'china', 'beijing', 'india']


---